REGRESSION (256782@main): Animating visibility with a display:contents child causes an element to disappear

anttijk opened the pull request:
#### 53814777b5f2c64071b4822a0a4898d516716492
<pre>
REGRESSION (256782@main): Animating visibility with a display:contents child causes an element to disappear
<a href="https://bugs.webkit.org/show_bug.cgi?id=251597">https://bugs.webkit.org/show_bug.cgi?id=251597</a>
rdar://105102892

Reviewed by Cameron McCormack.

We currently use ElementRareData::computedStyle() to store styles for Elements that don&apos;t have renderers
because they have &quot;display:contents&quot;. This creates various problems because the computedStyle field can
update outside normal style resolution/render tree update (as a performance optimization). This can lead
to substly inconsistent state where display:contents style doesn&apos;t match the render tree style and
some assumptions break.

This patch gives display:contents style a separate field in element rare data that is only written to
by RenderTreeUpdater. This way it is always in sync with the render tree style.

* LayoutTests/fast/animation/animation-display-contents-expected.html: Added.
* LayoutTests/fast/animation/animation-display-contents.html: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasDisplayContents const):
(WebCore::Element::storeDisplayContentsStyle):
(WebCore::Element::clearDisplayContentsStyle):
(WebCore::Element::existingComputedStyle const):
(WebCore::Element::renderOrDisplayContentsStyle const):
(WebCore::Element::resolveComputedStyle):
(WebCore::Element::resetComputedStyle):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::displayContentsStyle const):
(WebCore::ElementRareData::setDisplayContentsStyle):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updatePseudoElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::affectsRenderedSubtree):
(WebCore::Style::TreeResolver::resolveComposedTree):

Canonical link: <a href="https://commits.webkit.org/260270@main">https://commits.webkit.org/260270@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9f89fcdd6c293eeb28a82193f8255781b4c246a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16817 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40658 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116906 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111651 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8133 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99940 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113518 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/96965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/18231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/28606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/9783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10468 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/49538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7099 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12031 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -342,8 +342,10 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
+    else {
+        element.clearDisplayContentsStyle();
         element.resetComputedStyle();
+    }
```
Should we now be calling `resetComputedStyle()` regardless of `hasDisplayContents`? Since that's now used purely for storing `getComputedStyle` results, and we have a render object now.

anttijk replied:
You are right, resetComputedStyle here can be removed now.

heycam reviewed (approved):
r=me with the `resetComputedStyle()` change if that's correct.

anttijk committed:
REGRESSION (256782@main): Animating visibility with a display:contents child causes an element to disappear
https://bugs.webkit.org/show_bug.cgi?id=251597
rdar://105102892

Reviewed by Cameron McCormack.

We currently use ElementRareData::computedStyle() to store styles for Elements that don't have renderers
because they have "display:contents". This creates various problems because the computedStyle field can
update outside normal style resolution/render tree update (as a performance optimization). This can lead
to substly inconsistent state where display:contents style doesn't match the render tree style and
some assumptions break.

This patch gives display:contents style a separate field in element rare data that is only written to
by RenderTreeUpdater. This way it is always in sync with the render tree style.

* LayoutTests/fast/animation/animation-display-contents-expected.html: Added.
* LayoutTests/fast/animation/animation-display-contents.html: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasDisplayContents const):
(WebCore::Element::storeDisplayContentsStyle):
(WebCore::Element::clearDisplayContentsStyle):
(WebCore::Element::existingComputedStyle const):
(WebCore::Element::renderOrDisplayContentsStyle const):
(WebCore::Element::resolveComputedStyle):
(WebCore::Element::resetComputedStyle):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::displayContentsStyle const):
(WebCore::ElementRareData::setDisplayContentsStyle):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updatePseudoElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::affectsRenderedSubtree):
(WebCore::Style::TreeResolver::resolveComposedTree):

Canonical link: https://commits.webkit.org/260270@main
